cupy/_creation/matrix.py

```diff
@@ -141,7 +141,39 @@ def triu(m, k=0):
     return cupy.where(mask, m.dtype.type(0), m)
 
 
-# TODO(okuta): Implement vander
+def vander(x, N=None, increasing=False):
+    """Returns a Vandermonde matrix.
+
+    Args:
+        x (array-like): 1-D array or array-like object.
+        N (int, optional): Number of columns in the output.
+            ``N = len(x)`` by default.
+        increasing (bool, optional): Order of the powers of the columns.
+            If True, the powers increase from right to left,
+            if False (the default) they are reversed.
+
+    Returns:
+        cupy.ndarray: A Vandermonde matrix.
+
+    .. seealso:: :func:`numpy.vander`
+
+    """
+    x = cupy.asarray(x)
+    if x.ndim != 1:
+        raise ValueError("x must be a one-dimensional array or sequence.")
+    if N is None:
+        N = len(x)
+
+    v = cupy.empty((len(x), N), dtype=numpy.promote_types(x.dtype, int))
+    tmp = v[:, ::-1] if not increasing else v
+
+    if N > 0:
+        tmp[:, 0] = 1
+    if N > 1:
+        tmp[:, 1:] = x[:, None]
+        cupy.cumprod(tmp[:, 1:], out=tmp[:, 1:], axis=1)
```

Comment:
```suggestion
    cupy.power(x.reshape(-1, 1), cupy.arange(N), out=tmp)
```
How about using `cupy.power` instead for performance?

<details>

Benchmark script:
```py
import cupy
import cupyx.profiler

x = cupy.arange(1000, dtype='float32')

perf = cupyx.profiler.benchmark(cupy.vander, (x, 1000))
print(perf)
```

Result (original)
```
vander              :    CPU:  118.208 us   +/- 2.582 (min:  115.581 / max:  214.753) us     GPU-0:  208.787 us   +/-21.077 (min:  199.680 / max:  329.728) us
```

Result (suggestion):
```
vander              :    CPU:   49.705 us   +/-19.058 (min:   46.526 / max: 1923.795) us     GPU-0:   76.910 us   +/-20.474 (min:   64.512 / max: 1950.720) us
```

Reply:
That's cool! I'm surprised that the proposed impl is faster in cupy while cumulative prod is preferred in numpy( https://github.com/numpy/numpy/commit/30aeb2ee59d8b52bdd1ace33ac3d9b4e45c17b32) ...😯